File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/wallet/components/BalanceDisplay.kt
package app.getvela.wallet.feature.wallet.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import app.getvela.wallet.core.designsystem.components.VelaIcons
import app.getvela.wallet.core.designsystem.theme.VelaTheme
import app.getvela.wallet.core.designsystem.tokens.VelaFontFamily
import app.getvela.wallet.core.designsystem.tokens.VelaFontWeight
import app.getvela.wallet.core.designsystem.tokens.VelaIconSize
import app.getvela.wallet.core.designsystem.tokens.VelaSpacing
import app.getvela.wallet.core.designsystem.tokens.VelaTextSize
import app.getvela.wallet.feature.wallet.BalanceModel
import app.getvela.wallet.feature.wallet.BalanceStateKind

/**
 * Hero balance (spec vocabulary #4): label line (总余额 · USD), amount with
 * de-emphasised decimals, and exactly one of normal / zero-live / loading /
 * hidden, plus the optional BalanceStatusLine slot.
 */
@Composable
fun BalanceDisplay(
    model: BalanceModel,
    modifier: Modifier = Modifier,
    onToggleVisibility: () -> Unit = {},
    onStatusClick: () -> Unit = {},
) {
    val colors = VelaTheme.colors
    Column(modifier = modifier) {
        Text(
            text = "${model.label} · ${model.currency}",
            color = colors.fgSubtle,
            fontFamily = VelaFontFamily,
            fontWeight = VelaFontWeight.medium,
            fontSize = VelaTextSize.sm,
        )
        Spacer(modifier = Modifier.height(VelaSpacing.sm))
        when (model.state) {
            BalanceStateKind.Normal -> AmountRow(model, onToggleVisibility)
            BalanceStateKind.ZeroLive -> {
                AmountRow(model, onToggleVisibility)
                Spacer(modifier = Modifier.height(VelaSpacing.md))
                LiveIndicatorRow(model.liveText.orEmpty())
            }
            BalanceStateKind.Loading -> SkeletonBalanceBlock()
            BalanceStateKind.Hidden -> HiddenRow(model, onToggleVisibility)
        }
        model.status?.let { status ->
            Spacer(modifier = Modifier.height(VelaSpacing.md))
            BalanceStatusLine(model = status, onClick = onStatusClick)
        }
    }
}

@Composable
private fun AmountRow(model: BalanceModel, onToggle: () -> Unit = {}) {
    val colors = VelaTheme.colors
    // Spec 048: the amount itself hides the figures (the web's BalanceDisplay button).
    Row(modifier = Modifier.clickable(onClick = onToggle)) {
        Text(
            text = model.integer.orEmpty(),
            color = colors.fgBase,
            fontFamily = VelaFontFamily,
            fontWeight = VelaFontWeight.bold,
            fontSize = VelaTextSize.xl5,
            modifier = Modifier.alignByBaseline(),
        )
        model.decimals?.let { decimals ->
            Text(
                text = model.decimalMark + decimals,
                color = colors.fgSubtle,
                fontFamily = VelaFontFamily,
                fontWeight = VelaFontWeight.bold,
                fontSize = VelaTextSize.xl3,
                modifier = Modifier.alignByBaseline(),
            )
        }
    }
}

@Composable
private fun LiveIndicatorRow(text: String) {
    val colors = VelaTheme.colors
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(VelaSpacing.md),
    ) {
        Box(
            modifier = Modifier
                .size(WalletMetrics.liveDotSize)
                .alpha(rememberPulseAlpha())
                .background(colors.successBase, CircleShape),
        )
        Text(
            text = text,
            color = colors.fgMuted,
            fontFamily = VelaFontFamily,
            fontWeight = VelaFontWeight.medium,
            fontSize = VelaTextSize.sm,
        )
    }
}

@Composable
private fun HiddenRow(model: BalanceModel, onToggleVisibility: () -> Unit) {
    val colors = VelaTheme.colors
    Row(
        modifier = Modifier.clickable(onClick = onToggleVisibility),verticalAlignment = Alignment.CenterVertically) {
        Text(
            text = model.integer.orEmpty(),
            color = colors.fgBase,
            fontFamily = VelaFontFamily,
            fontWeight = VelaFontWeight.bold,
            fontSize = VelaTextSize.xl4,
        )
        Spacer(modifier = Modifier.width(VelaSpacing.md))
        IconButton(onClick = onToggleVisibility) {
            Icon(
                imageVector = VelaIcons.EyeOff,
                // While hidden, the affordance reveals: announce "show balance".
                contentDescription = model.a11yShow,
                tint = colors.fgMuted,
                modifier = Modifier.size(VelaIconSize.lg),
            )
        }
    }
}

File: app-ios/VelaWallet/VelaWallet/Components/Wallet/BalanceDisplay.swift
//
//  BalanceDisplay.swift
//  VelaWallet
//
//  BalanceDisplay + BalanceStatusLine (spec 015 vocabulary #4/#5): label
//  line (总余额 · USD), hero amount with de-emphasised decimals, exactly
//  one of normal / zero-live (pulsing dot + 实时·监听收款中) / loading
//  (skeleton block) / hidden (six dots + eye-off), plus an optional
//  warning/refreshing status line.
//

import SwiftUI

struct BalanceDisplay: View {
    @Environment(\.theme) private var theme
    @Environment(\.walletTextScale) private var textScale

    let model: BalanceModel
    /// Where the status line goes. Absent in the gallery, where the line is a
    /// picture of a state rather than a way out of one.
    var onStatusTap: (() -> Void)?

    var body: some View {
        VStack(alignment: .leading, spacing: Tokens.Space.s8) {
            Text(verbatim: "\(model.label) · \(model.currency)")
                .typeRole(Typography.label.scaled(textScale))
                .foregroundStyle(theme.fgMuted)
            amount
            if let live = model.liveText {
                liveRow(live)
            }
            if let status = model.status {
                statusDoor(status)
            }
        }
    }

    @ViewBuilder private var amount: some View {
        switch model.state {
        case .loading:
            SkeletonBlock(width: WalletGeometry.skeletonBalanceWidth, height: WalletGeometry.skeletonBalanceHeight)
        case .hidden:
            HStack(spacing: Tokens.Space.s16) {
                HStack(spacing: Tokens.Space.s8) {
                    ForEach(0..<WalletGeometry.hiddenDotCount, id: \.self) { _ in
                        Circle()
                            .fill(theme.fgBase)
                            .frame(width: WalletGeometry.hiddenDot, height: WalletGeometry.hiddenDot)
                    }
                }
                LucideIcon(.eyeOff, size: LucideIconSize.eye)
                    .foregroundStyle(theme.fgMuted)
            }
            .frame(minHeight: WalletGeometry.skeletonBalanceHeight)
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(Text(verbatim: model.a11yShow))
        case .normal, .zeroLive:
            (Text(verbatim: model.integer ?? "")
                .font(Typography.amountHero.scaled(textScale).font)
                .foregroundStyle(theme.fgBase)
                + Text(verbatim: model.decimals.map { ".\($0)" } ?? "")
                .font(Typography.amountHeroDecimals.scaled(textScale).font)
                .foregroundStyle(theme.fgMuted))
                .lineLimit(1)
                .minimumScaleFactor(WalletGeometry.heroMinScale)
                .accessibilityLabel(Text(verbatim: "\(model.integer ?? "").\(model.decimals ?? "") \(model.currency)"))
        }
    }

    private func liveRow(_ text: String) -> some View {
        HStack(spacing: Tokens.Space.s8) {
            PulsingDot(color: theme.successBase)
            Text(verbatim: text)
                .typeRole(Typography.rowSub.scaled(textScale))
                .foregroundStyle(theme.successBase)
        }
    }

    /// The line, as a control where there is somewhere to go.
    ///
    /// It has always drawn a chevron — a promise of a destination — and had
    /// none behind it (row 12 of 057's audit). Android and the web agree on
    /// what it opens: a failed chain's RPC fix, anything else the breakdown.
    @ViewBuilder private func statusDoor(_ status: BalanceStatusModel) -> some View {
        if let onStatusTap {
            Button(action: onStatusTap) {
                statusRow(status).contentShape(Rectangle())
            }
            .buttonStyle(.plain)
        } else {
            statusRow(status)
        }
    }

    private func statusRow(_ status: BalanceStatusModel) -> some View {
        let tint = status.kind == .warning ? theme.warningBase : theme.fgMuted
        return HStack(spacing: Tokens.Space.s8) {
            LucideIcon(status.kind == .warning ? .triangleAlert : .refreshCw, size: LucideIconSize.statusIcon)
                .foregroundStyle(tint)
            Text(verbatim: status.text)
                .typeRole(Typography.rowSub.scaled(textScale))
                .foregroundStyle(tint)
            LucideIcon(.chevronRight, size: LucideIconSize.smallChevron)
                .foregroundStyle(theme.fgSubtle)
        }
    }
}

/// Zero-live indicator dot — opacity pulse, static under Reduce Motion.
private struct PulsingDot: View {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var dimmed = false
    let color: Color

    var body: some View {
        Circle()
            .fill(color)
            .frame(width: WalletGeometry.liveDot, height: WalletGeometry.liveDot)
            .opacity(dimmed && !reduceMotion ? Tokens.Opacity.dim : 1)
            .animation(
                reduceMotion ? nil : .easeInOut(duration: Tokens.Motion.slow).repeatForever(autoreverses: true),
                value: dimmed
            )
            .onAppear { dimmed = true }
    }
}

#Preview("Balance states dark") {
    VStack(alignment: .leading, spacing: Tokens.Space.s24) {
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .normal,
            integer: "$1,383", decimals: "28", liveText: nil, status: nil,
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .zeroLive,
            integer: "$0", decimals: "00", liveText: "实时 · 监听收款中", status: nil,
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .loading,
            integer: nil, decimals: nil, liveText: nil, status: nil,
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .hidden,
            integer: "••••••", decimals: nil, liveText: nil, status: nil,
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .normal,
            integer: "$1,383", decimals: "46", liveText: nil,
            status: BalanceStatusModel(kind: .warning, text: "部分代币无法获取价格。"),
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
        BalanceDisplay(model: BalanceModel(
            label: "总余额", currency: "USD", state: .normal,
            integer: "$1,383", decimals: "28", liveText: nil,
            status: BalanceStatusModel(kind: .refreshing, text: "部分余额仍在更新。"),
            a11yHide: "隐藏余额", a11yShow: "显示余额"
        ))
    }
    .padding(Tokens.Space.s24)
    .frame(maxWidth: .infinity, alignment: .leading)
    .background(Tokens.dark.bgBase.color)
    .themed(.dark)
}

#Preview("Balance light") {
    BalanceDisplay(model: BalanceModel(
        label: "Total balance", currency: "USD", state: .normal,
        integer: "$1,383", decimals: "28", liveText: nil, status: nil,
        a11yHide: "Hide balance", a11yShow: "Show balance"
    ))
    .padding(Tokens.Space.s24)
    .themed(.light)
}
